ext/drstatecmp/drstatecmp.dox

```diff
@@ -0,0 +1,146 @@
+
+There are two types of checks depending on whether a basic block has any side-effects.
+
+\subsection sec_drstatecmp_checks_side_effects_free Side-effect-free Basic Blocks
+
+Side-effect-free basic blocks are duplicated and executed twice, once with the full
+instrumentation sequence and once with no instrumentation except for a state comparison at
+the end. Essentially, we check whether the machine state at the end of the basic block is
+the same with and without instrumentation code. This check should catch any state
+clobbering by instrumentation.
+
+The side-effect-free basic blocks are duplicated at drmgr's app2app stage. The copy of
+each such basic block is appended immediately after the original basic block. Labels are
+inserted in the beginning of the original basic block (ORIG_BB) and its copy (COPY_BB) to
+allow identification in later stages. The terminating instruction of the basic block (if
+any) is only kept in the copy and is preceded by a TERM label. A jump to the TERM label is
+inserted at the end the original basic block to skip the duplicated bb.
+
+The insert instrumentation stage inserts the actual checks. First, the original basic
+block is instrumented. In the beginning of the original basic block, the machine state is
+saved. This state will be used for state restoration before the re-execution of the basic
+block. At the end of the original block, code is inserted to save the state for later
+comparison and restore the machine state to the state before executing this bb. This stage
+also removes the previously inserted jump instr to the TERM label and let it fall through
+to the instructions of the copy of the basic block. Then, the copy of the basic block is
+instrumented. At the end of the copy, a clean call is inserted that compares the current
+machine state with the machine state that was saved at the end of the original basic
+block. Finally, all previously-inserted (from other passes) instrumentation instructions
+are removed from the basic block copy, making it instrumentation-free and enabling
+identification of various types of instrumentation clobering.
+
+\subsection sec_drstatecmp_checks_side_effects Basic Blocks with Side Effects
+
+Checking basic blocks with side effects is not yet implemented but this section serves
+as a high-level design overview.
+
+Basic blocks with side-effects cannot be executed twice (at least not without a lot of
+extra complexity to monitor and restore memory state). In this case, the checks are
+inserted throughout the basic block whenever needed taking into account the app
+instructions, existing instrumentation, and the lazy restoration by drreg. As a result,
+incomplete-decoder bugs and lazy-condition-ddreg bugs will not be detected.
+
+Instructions with side effects include instructions that write to memory, interrupts,
+and system calls. To avoid inter-procedural analysis, all function calls are
+conservatively assumed to have side effects.
```

Comment:
I think this last sentence can be removed (function calls are are not embedded; they'll become separate blocks for each piece)